pkg/common/x509util/dns.go

```diff
@@ -3,48 +3,106 @@ package x509util
 import (
 	"errors"
 	"fmt"
-	"regexp"
 	"strings"
+
+	"golang.org/x/exp/utf8string"
+	"golang.org/x/net/idna"
 )
 
-var isDNSLabel = regexp.MustCompile(`^[a-zA-Z0-9]([-]*[a-zA-Z0-9])+$`).MatchString
+var (
+	ErrTooManyWildcards         = errors.New("too many wildcards")
+	ErrWildcardMustBeFirstLabel = errors.New("wildcard must be first label")
+	ErrEmptyDomain              = errors.New("empty domain")
+	ErrIDNAError                = errors.New("idna error")
+	ErrDomainEndsWithDot        = errors.New("domain ends with dot")
+	ErrWildcardOverlap          = errors.New("wildcard overlap")
+	ErrNameMustBeASCII          = errors.New("name must be ascii")
+	ErrLabelMismatchAfterIDNA   = errors.New("label mismatch after idna")
+	errNoWildcardAllowed        = errors.New("wildcard not allowed")
+)
 
-// ValidateDNS validates that provided string is a valid DNS name
-func ValidateDNS(dns string) error {
-	// follow https://tools.ietf.org/html/rfc5280#section-4.2.1.6
-	// do not allow empty or the technically valid DNS " "
-	dns = strings.TrimSpace(dns)
-	if len(dns) == 0 {
-		return errors.New("empty or only whitespace")
+func ValidateAndNormalize(domain string) (string, error) {
```

Comment:
This function no longer normalizes. We should probably take that out of the name and not return a string.